beacon-chain/verification/cache.go

```diff
@@ -0,0 +1,196 @@
+	// SignatureVerified accesses the result of a previous signature verification.
+	SignatureVerified(sig SignatureData) (bool, error)
+}
+
+// SignatureData represents the set of parameters that together uniquely identify a signature observed on
+// a beacon block. This is used as the key for the signature cache.
+type SignatureData struct {
+	Root      [32]byte
+	Parent    [32]byte
+	Signature [96]byte
+	Proposer  primitives.ValidatorIndex
+	Slot      primitives.Slot
+}
+
+func (d SignatureData) logFields() log.Fields {
+	return log.Fields{
+		"root":        fmt.Sprintf("%#x", d.Root),
+		"parent_root": fmt.Sprintf("%#x", d.Parent),
+		"signature":   fmt.Sprintf("%#x", d.Signature),
+		"proposer":    d.Proposer,
+		"slot":        d.Slot,
+	}
+}
+
+func newSigCache(vr []byte, size int) *sigCache {
+	return &sigCache{Cache: lruwrpr.New(size), valRoot: vr}
+}
+
+type sigCache struct {
+	*lru.Cache
+	valRoot []byte
+}
+
+// VerifySignatureWithState verifies the given signature data against the key obtained via ValidatorAtIndexer.
+func (c *sigCache) VerifySignature(sig SignatureData, v ValidatorAtIndexer) (err error) {
+	defer func() {
+		if err == nil {
+			c.Add(sig, true)
+		} else {
+			log.WithError(err).WithFields(sig.logFields()).Debug("caching failed signature verification result")
+			c.Add(sig, false)
+		}
+	}()
+	e := slots.ToEpoch(sig.Slot)
+	fork, err := forks.Fork(e)
+	if err != nil {
+		return err
+	}
+	domain, err := signing.Domain(fork, e, params.BeaconConfig().DomainBeaconProposer, c.valRoot)
+	if err != nil {
+		return err
+	}
+	pv, err := v.ValidatorAtIndex(sig.Proposer)
+	if err != nil {
+		return err
+	}
+	pb, err := bls.PublicKeyFromBytes(pv.PublicKey)
+	if err != nil {
+		return err
+	}
+	s, err := bls.SignatureFromBytes(sig.Signature[:])
+	if err != nil {
+		return err
+	}
+	sr, err := signing.ComputeSigningRootForRoot(sig.Root, domain)
+	if err != nil {
+		return err
+	}
+	if !s.Verify(pb, sr[:]) {
+		return signing.ErrSigFailedToVerify
+	}
+
+	return nil
+}
+
+// SignatureVerified checks the signature cache for the given key, and returns a boolean value of true
+// if it has been seen before, and an error value indicating whether the signature verification succeeded.
+// ie only a result of (true, nil) means a previous signature check passed.
+func (c *sigCache) SignatureVerified(sig SignatureData) (bool, error) {
+	val, seen := c.Get(sig)
+	if !seen {
+		return false, nil
+	}
+	verified, ok := val.(bool)
+	if !ok {
+		log.WithFields(sig.logFields()).Debug("ignoring invalid value found in signature cache")
+		// This shouldn't happen, and if it does, the caller should treat it as a cache miss and run verification
+		// again to correctly populate the cache key.
+		return false, nil
+	}
+	if verified {
+		return true, nil
+	}
+	return true, signing.ErrSigFailedToVerify
+}
+
+// ProposerCache represents a type that can compute the proposer for a given slot + parent root,
```

Comment:
Why do we have a new proposer cache here ? We already have other proposer id caches that are already used all across the repo. Risk with using one here, is that you recompute shuffling twice

Reply:
This is an improvement on the current state of affairs where we currently recompute shuffling for every gossip verification. That said we want to consolidate proposer id caches so we don't wind up with 4 of them, I am working with @potuz on a strategy.

Reply:
potuz is working on his proposer cache improvements. In the meantime I've changed this pr to stub the cache, returning false for a cache lookup and so always using full stategen+proposer calculation. This is equivalent to what the current sync code does, but leaves the pieces in place to easily plug in the shared cache once its ready.